File: DP/188. Best Time to Buy and Sell Stock IV.cpp
/*


188. Best Time to Buy and Sell Stock IV


Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete at most k transactions.

Note:
You may not engage in multiple transactions at the same time (ie, you must sell the stock before you buy again).



*/


class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        if(k>prices.size()/2)
        {
            int profit = 0;
            for(int i = 1; i<prices.size();i++){
                profit += max(prices[i]-prices[i-1],0);
            }
            return profit;
        }
        
        if(prices.size()==0) return 0;
        vector<int>dp(prices.size(),0);
        
        for(int n = 1; n<=k && n<prices.size(); n++){
            int tempmax = -prices[0]; //tempmax 相当于negative cost，最小化cost 相当于最大化负的cost
            for(int i = 1; i<prices.size();i++){
                int lastprof = dp[i];
                dp[i] = max(tempmax + prices[i], dp[i-1]);
                tempmax = max(tempmax, lastprof-prices[i]);
            }
        }
        return dp[prices.size()-1];
    }
};

class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        if(prices.size()<=1) return 0;
        if(k>prices.size()/2){
            int profit = 0;
            for(int i = 1; i<prices.size(); i++)
                profit += max(prices[i]-prices[i-1],0);
            return profit;
        }
        vector<int>dp(prices.size(),0);
        for(int a = 0; a<k; a++){
            int cost = prices[0], prev_cost = prices[0];
            for(int i = 1; i<prices.size(); prev_cost = cost, i++){
                cost = min(prices[i]-dp[i],cost); 
                dp[i] = max(prices[i]-prev_cost,dp[i-1]);
            }
        }
        return dp[prices.size()-1];
    }
};


//not a good solution but it works
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        if(prices.size()==0) return 0;
        vector<int>dp(prices.size(),0);
        
        for(int n = 1; n<=k && n<prices.size(); n++){
            int tempmax = -prices[0];
            for(int i = 1; i<prices.size();i++){
                int lastprof = dp[i];
                dp[i] = max(tempmax + prices[i], dp[i-1]);
                tempmax = max(tempmax, lastprof-prices[i]);
            }
        }
        return dp[prices.size()-1];
    }
};

File: Dynamic Programming/188. Best Time to Buy and Sell Stock IV.py
"""
188. Best Time to Buy and Sell Stock IV
"""

import functools
class Solution:
    def maxProfit(self, k, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        if len(prices)<=1: return 0
        if k>=len(prices)/2 :
            return functools.reduce(lambda x,i: x+max(prices[i]-prices[i-1],0),range(1,len(prices)),0)
        profit = [0 for _ in range(len(prices))]
        for _ in range(k):
            cost= prices[0]
            for i, p in enumerate(prices[1:]):
                cost, profit[i+1] = min(p - profit[i+1],cost), max(profit[i],p-cost)
        return profit[-1]
        

import functools
class Solution:
    def maxProfit(self, k, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        if len(prices)<=1 or k==0: return 0
        if k>=len(prices)/2 :
            return functools.reduce(lambda x,i: x+max(prices[i]-prices[i-1],0),range(1,len(prices)),0)
        buy = [prices[0] for _ in range(k)]
        sell = [0 for _ in range(k)]
        for _ , p in enumerate(prices[1:]):
            for i in range(k):
                buy[i] = min(p, buy[i]) if i == 0 else min(p-sell[i-1], buy[i]) 
                sell[i] = max(p-buy[i],sell[i])
            #print(buy, sell)
        return sell[-1]
        